Add support for wlr-foreign-toplevel-management

ftyghome opened the pull request:
This PR introduces support for the **wlr-foreign-toplevel-management** protocol in Cage.

The protocol is especially useful in kiosk scenarios. In our use case, the system needs to run multiple windows (for example, a VNC remote connection window and a WebKit browser). We would like these windows to remain active simultaneously and allow a background daemon to handle requests to switch between them. This enables smoother workflows where the remote connection window and the browser can be freely toggled when messages arrive.

I also noticed related discussions in the following issues, where adding support for this protocol has been requested:
	•	https://github.com/cage-kiosk/cage/issues/99
	•	https://github.com/cage-kiosk/cage/issues/302
	•	https://github.com/cage-kiosk/cage/issues/391

This implementation aims to address those needs by integrating the protocol directly into Cage.

emersion commented on xdg_shell.c:
```diff
@@ -240,6 +294,19 @@ handle_xdg_toplevel_destroy(struct wl_listener *listener, void *data)
+		xdg_shell_view->foreign_toplevel = NULL;
+	}
+
+	if (view->foreign_toplevel_handle) {
```
This condition is unnecessary: `wlr_foreign_toplevel_handle_v1_destroy()` is a no-op if NULL is passed in (like `free()`).

emersion reviewed (commented):
It seems like support for Xwayland is missing?

emersion commented on xwayland.c:
```diff
@@ -176,6 +192,27 @@ handle_xwayland_dissociate(struct wl_listener *listener, void *data)
 }
 
+static void
+handle_xwayland_surface_request_activate(struct wl_listener *listener, void *data)
```
This is the same code as xdg-shell. Can it be moved to `view.c`?

ftyghome commented:
Thanks for the review! I've moved the foreign-toplevel creation and destruction logic to the map/unmap paths, and moved the duplicated code to view.c. Please let me know if there's anything else needed:)

emersion commented on xwayland.c:
```diff
@@ -141,6 +150,9 @@ handle_xwayland_surface_destroy(struct wl_listener *listener, void *data)
 	wl_list_remove(&xwayland_view->destroy.link);
 	wl_list_remove(&xwayland_view->request_fullscreen.link);
+	wl_list_remove(&xwayland_view->associate.link);
+	wl_list_remove(&xwayland_view->dissociate.link);
```
Ah, good catch! I noticed the same missing calls in https://github.com/cage-kiosk/cage/pull/459. Could we keep this bugfix separate from support for a new protocol?

emersion commented on view.c:
```diff
@@ -144,6 +181,17 @@ view_map(struct cg_view *view, struct wlr_surface *surface)
+
+	view->foreign_toplevel_handle = wlr_foreign_toplevel_handle_v1_create(view->server->foreign_toplevel_manager);
+	if (!view->foreign_toplevel_handle) {
+		wlr_log(WLR_ERROR, "Failed to create foreign toplevel handle");
```
I think we can disconnect the client with `wl_resource_post_no_memory()` like is done above. This would avoid ending up with a half-constructed view.

We should make sure to `return` here to avoid accessing a NULL pointer below.

emersion commented on xdg_shell.h:
```diff
@@ -34,6 +33,7 @@ struct cg_xdg_popup {
 	struct wl_listener reposition;
 };
 
+struct cg_xdg_shell_view *xdg_shell_view_from_view(struct cg_view *view);
```
Once the comment above is resolved, this no longer needs to be declared in this header

emersion commented on view.c:
```diff
@@ -13,13 +13,16 @@
 #include "server.h"
 #include "view.h"
+
+#include "xdg_shell.h"
```
Ditto
